### ISSUE TITLE
Mismatched types with concat_slices macro whilst using char without initializing value.

### ISSUE THREAD

**Issue by airblast-dev** [rossmacarthur/constcat#2]
Whilst trying to concatenate two slices of chars using concat_slices I noticed rust-analyzer warning me about mismatched types saying "mismatched types expected char, found u8".  I am using rustc 1.75.0-nightly (960754090 2023-10-06).

### Example:
```rust
use constcat::concat_slices;
fn main() {
    const A_AND_B: [char;2] = ['a', 'b'];
    const SECRET_NUMS: [char; 2]= ['3', '4'];
    const MY_CONCAT_SLICE: &[char] = concat_slices!([char]: &A_AND_B, &SECRET_NUMS); // mismatched types expected char, found u8
}
```
Modifying the default internal initializer field in the macro with a char instead of an int or float fixed the issue. I am not familiar with macro's and their syntax. Here is the version accepting chars but not integers.

```rust
macro_rules! concat_slices {
    ([$init:expr; $T:ty]: $($s:expr),+ $(,)?) => {{
        $(
            const _: &[$T] = $s; // require constants
        )*
        const LEN: usize = $( $s.len() + )* 0;
        const ARR: [$T; LEN] = {
            let mut arr: [$T; LEN] = [$init; LEN];
            let mut base: usize = 0;
            $({
                let mut i = 0;
                while i < $s.len() {
                    arr[base + i] = $s[i];
                    i += 1;
                }
                base += $s.len();
            })*
            if base != LEN { panic!("invalid length"); }
            arr
        };
        &ARR
    }};

    ([$T:ty]: $($s:expr),+ $(,)?) => {
        concat_slices!(['\x00'; $T]: $($s),+)
    };
}

fn main() {
    const A_AND_B: [char;2] = ['a', 'b'];
    const SECRET_NUMS: [char; 2]= ['3', '4'];
    const MY_CONCAT_SLICE: &[char]= concat_slices!([char]: &A_AND_B, &SECRET_NUMS);
    const MY_CONCAT_SLICEA: &[usize]= concat_slices!([char]: &[3,4], &[1,2]); // ERROR: expected char, found integer.
}
```
I am happy to learn a bit more of the macro syntax and make a commit if you are busy or whatever :) .

EDIT: Reading the docs again, the examples use an initializer and the doc mentions optionally providing an initializer for non integer types such as chars and floats. Idk if it's a typo in the docs, or I'm just missing something.

**Comment by rossmacarthur** [comment]
Hi, you can provide a `'\x00'` initializer for char slices like this

```rust
use constcat::concat_slices;
fn main() {
    const A_AND_B: [char;2] = ['a', 'b'];
    const SECRET_NUMS: [char; 2]= ['3', '4'];
    const MY_CONCAT_SLICE: &[char] = concat_slices!(['\x00'; char]: &A_AND_B, &SECRET_NUMS);
}
```

**Comment by rossmacarthur** [comment]
Theres actually no good reason why these can't be special cased, I have made a change in 66a2927a38270d28ac0246c8699f21f93573543c that allows `[char]`, `[f32]` and `[f64]` slices to work without the initializer specified. This is released in version 0.3.1.

**Comment by airblast-dev** [comment]
Awesome! It's just that the docs mentioned optionally providing one for chars and floats, so it got me confused. After looking at the source code, I applied the solution you posted above. Thanks for your time and the update :)
